modules/ballerina-builtin/src/main/ballerina/ballerina/caching/natives.bal

```diff
@@ -0,0 +1,201 @@
+        function (error) onErrorFunction = null;
+        error schedulerError;
+        scheduleID, schedulerError = task:scheduleTimer(onTriggerFunction, onErrorFunction, {delay:CACHE_CLEANUP_START_DELAY, interval:CACHE_CLEANUP_INTERVAL});
+        // If task creation failed, throw an error.
+        if (schedulerError != null) {
+            throw schedulerError;
+        }
+    }
+
+    // Create a new cache.
+    cache c = {name:name, timeOut:timeOut, capacity:capacity, evictionFactor:evictionFactor, entries:{}};
+    // Get the current total cache count.
+    int currentCachesCount = lengthof caches;
+    // Add the new cache to the end of the caches array.
+    caches[currentCachesCount] = c;
+    // Return the new cache.
+    return c;
+}
+
+@doc:Description {value:"Adds the given key, value pair to the provided cache."}
+@doc:Param {value:"cache - a cache"}
+@doc:Param {value:"key - value which should be used as the key"}
+@doc:Param {value:"value - value to be cached"}
+public function put (cache c, string key, any value) {
+    int maxCapacity = c.capacity;
+    int currentCapacity = c.entries.length();
+    // if the current cache is full,
+    if (maxCapacity <= currentCapacity) {
+        evictCache(c);
+    }
+    // Add the new entry
+    int time = currentTime().time;
+    cacheEntry entry = {value:value, lastAccessedTime:time};
+    c.entries[key] = entry;
+}
+
+function evictCache (cache c) {
+    int maxCapacity = c.capacity;
+    float evictionFactor = c.evictionFactor;
+    int noOfEntriesToBeEvicted = <int>(maxCapacity * evictionFactor);
+    int i = 0;
+    while (i < noOfEntriesToBeEvicted) {
+        string cacheKey = getLRUCache(c);
+        c.entries.remove(cacheKey);
+        i = i + 1;
+    }
+}
+
+@doc:Description {value:"Returns the cached value associated with the given key. Returns null if the provided key does not exist in the cache."}
+@doc:Param {value:"cache - a cache"}
+@doc:Param {value:"key - key which is used to retrieve the cached value"}
+public function get (cache c, string key) returns (any) {
+    any value = c.entries[key];
+    if (value == null) {
+        return null;
+    }
+    var entry, e = (cacheEntry)value;
+    if (e != null || entry == null) {
+        return null;
+    }
+    entry.lastAccessedTime = currentTime().time;
+    return entry.value;
+}
+
+@doc:Description {value:"Removes a cached value from a cache."}
+@doc:Param {value:"cache - a cache"}
+@doc:Param {value:"key - key of the cache entry which needs to be removed"}
+public function remove (cache c, string key) {
+    c.entries.remove(key);
+}
+
+@doc:Description {value:"Clears a cache."}
+public function cleanCache () returns (error) {
+    int currentCacheIndex = 0;
+    int cacheSize = lengthof caches;
+    // Iterate through all caches.
+    while (currentCacheIndex < cacheSize) {
+        // Get a cache from the array.
+        cache currentCache = caches[currentCacheIndex];
```

Comment:
There is a bug in your code. You are creating a timer task per cache and each of those tasks is running cache expiry for all the caches in the system. Ideally you would create one timer task to scan all caches and expire the entries.